### PR TITLE
bugfix: array literals are array_exprt, not constant_exprt

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-programs/interpreter.cpp [hunk 1]
```diff
@@ -551,7 +551,7 @@ exprt interpretert::get_value(
   }
   else if(real_type.id()==ID_array)
   {
-    constant_exprt result(type);
+    array_exprt result(to_array_type(real_type));
```
Comment (tautschnig) [inline src/goto-programs/interpreter.cpp:554]: Is this code being tested somewhere?

File: src/goto-programs/interpreter.cpp [hunk 2]
```diff
@@ -603,12 +603,7 @@ exprt interpretert::get_value(
   else if(real_type.id() == ID_pointer)
   {
     if(rhs[integer2size_t(offset)]==0)
-    {
-      // NULL pointer
-      constant_exprt result(type);
-      result.set_value(ID_NULL);
-      return result;
-    }
+      return null_pointer_exprt(to_pointer_type(real_type)); // NULL pointer
```
Comment (tautschnig) [inline src/goto-programs/interpreter.cpp:606]: Unrelated to commit subject?

File: src/goto-programs/interpreter.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-programs/interpreter.cpp:554]: I doubt it -- there are no tests for the interpreter, I believe.

File: src/goto-programs/interpreter.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-programs/interpreter.cpp:606]: Indeed, will split.

File: src/goto-programs/interpreter.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-programs/interpreter.cpp:554]: Not in this repository, but I believe that's happening in TG or some other private repo - @peterschrammel?

**Review by tautschnig — COMMENTED** [review]
Looks good to me, but I'd appreciate if someone with insight into TG and other tools could do the actual approval.

File: src/goto-programs/interpreter.cpp [hunk 3]
```diff
@@ -551,7 +551,7 @@ exprt interpretert::get_value(
   }
   else if(real_type.id()==ID_array)
   {
-    constant_exprt result(type);
+    array_exprt result(to_array_typet(real_type));
```
Comment (tautschnig) [inline src/goto-programs/interpreter.cpp]: `to_array_type`, not `to_array_typet`

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: ff01bc165760de9008a272df5a3699d866a2db93).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85605139

**Review by peterschrammel — APPROVED** [review]
Looks fine. Also wrt TG.
